### PR TITLE
RavenDB-21821: Routing performance improvements.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21821

### Additional description
Routing is performance critical. This PR improves a few aspects of routing allowing less allocations and faster acquisition of the required resources. 

### Type of change
- Optimization

### How risky is the change?
- Moderate

### Backward compatibility
- Non breaking change

### CODE REVIEW

**Comment by redknightlois** [comment]
image

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 1]
```diff
@@ -660,18 +660,11 @@ public bool TryGetDatabaseIfLoaded(StringSegment databaseName, out DocumentDatab
             return task.Result.DatabaseShutdown.IsCancellationRequested == false;
         }
 
-        public sealed class DatabaseSearchResult
+        public struct DatabaseSearchResult(DatabaseSearchResult.Status databaseStatus, Task<DocumentDatabase> databaseTask, ShardedDatabaseContext databaseContext)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: readonly struct?

Comment (redknightlois) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: Semantically correct. Yes

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 2]
```diff
@@ -11,22 +12,30 @@
 using Raven.Client.Util;
 using Sparrow;
 using Sparrow.Collections;
+using static Elastic.Clients.Elasticsearch.JoinField;
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: Please remove it

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 3]
```diff
@@ -86,44 +101,58 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
...
-        private void RemoveCaseSensitive(StringSegment resourceName, Task<TResource> resourceTask)
+
+        private void RemoveCaseSensitive(StringSegment resourceName)
         {
             if (_mappings.TryGetValue(resourceName, out ConcurrentSet<StringSegment> mappings))
             {
                 foreach (var mapping in mappings)
                 {
-                    _caseSensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(mapping, resourceTask));
+                    _caseSensitive.Remove(mapping, out _);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: You must not do this. It's super important that we remove only if both elements from KeyValyePair are matching. See details here:

https://github.com/ravendb/ravendb/pull/14602

It's likely the cause of failing test: https://jenkins.hibernatingrhinos.com/job/PR_Tests/9494/testReport/junit/SlowTests.Issues/RavenDB_19002/ResourceCacheMustNotAllowToLeakDatabaseInstance/

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 4]
```diff
@@ -86,44 +101,58 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
...
             }
 
             return true;
         }
 
-        public bool TryGetAndRemove(StringSegment resourceName, out Task<TResource> resourceTask)
+        private bool TryRemoveInternal(StringSegment resourceName, Task<TResource> resourceTask)
         {
-            if (TryGetValue(resourceName, out resourceTask) == false)
+            if (_caseInsensitive.Remove(resourceName, out _) == false)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: This is wrong. Same issue as below. It must stay as it was:
```
_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)
```

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 5]
```diff
@@ -86,44 +101,58 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
...
         {
-            if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
-                return false;
-
-            _resourceDetails.TryRemove(resourceTask, out _);
-
             lock (this)
             {
-                RemoveCaseSensitive(resourceName, resourceTask);
+                if (TryRemoveInternal(resourceName, resourceTask) == false)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: Why this is under the lock now?

Comment (arekpalinski) [reply src/Raven.Server/Documents/ResourceCache.cs]: What is the reason of introducing `TryRemoveInternal()`? It makes that it's more difficult to track the changes. This is quite sensitive piece of code. I'd prefer to apply only the necessary changes without doing any refactoring if there is no good reason for that.

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 6]
```diff
@@ -133,46 +162,56 @@ IEnumerator IEnumerable.GetEnumerator()
...
+
+                _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
+
+                return value;
+            }
+        }
+
+        private Task<TResource> GetOrAddInternal(StringSegment databaseName, Task<TResource> task)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: Can we not introduce this method? It's just a refactoring and there is no other called of this piece of code. Same motivation as my above comment about `TryRemoveInternal()`

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 2 above)
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]: Why is that here?

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 7]
```diff
@@ -228,9 +271,15 @@ public IDisposable RemoveLockAndReturn(string databaseName, Action<TResource> on
                 {
                     _caseInsensitive.TryUpdate(databaseName, resourceLocked, current);
                     _resourceDetails.TryRemove(current, out _);
-                    RemoveCaseSensitive(databaseName, current);
+                    RemoveCaseSensitive(databaseName);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs]: `current` must be passed

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 3 above)
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]: Pretty sure that this is a problem - note that we were previously checking that both the key & value matched.
IIRC, that was a bug fix, see:

https://github.com/ravendb/ravendb/commit/34c9fcb5a111b352795f30ce24bd91d7a68bfe31

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 4 above)
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]:
```suggestion
            Debug.Assert(Monitor.IsEntered(this));
            if (_caseInsensitive.Remove(resourceName, out _) == false)
```

Comment (ayende) [reply src/Raven.Server/Documents/ResourceCache.cs]: Let's ensure that this invariant is held

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 8]
```diff
@@ -133,46 +162,56 @@ IEnumerator IEnumerable.GetEnumerator()
...
+                _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
+
+                return value;
+            }
+        }
+
+        private Task<TResource> GetOrAddInternal(StringSegment databaseName, Task<TResource> task)
+        {
```
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]:
```suggestion
        {
                Debug.Assert(Monitor.IsEntered(this));
```

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 9]
```diff
@@ -228,9 +271,15 @@ public IDisposable RemoveLockAndReturn(string databaseName, Action<TResource> on
                 {
                     _caseInsensitive.TryUpdate(databaseName, resourceLocked, current);
                     _resourceDetails.TryRemove(current, out _);
-                    RemoveCaseSensitive(databaseName, current);
+                    RemoveCaseSensitive(databaseName);
+
+                    // We need to refresh all of them.
+                    _readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                    _readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                    _readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
```
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]: That opens us up to race conditions, no? Shouldn't we do this under the lock (with the understanding that it can be expensive)?

**Review by arekpalinski — COMMENTED** [review]
1. I like the idea of the usage of frozen dictionaries but currently there are too many unnecessary changes which are introducing issues.

2.
   > How risky is the change?
   >    Low

   At least "Moderate" IMO. It's super hard to spot a possible concurrency issue here.

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 9 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ResourceCache.cs]: This is still under the lock so it's fine

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 2 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: I hate Resharper. Grrr

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 3 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: That's a nasty, not clear behavior. That deviation should be documented.

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 5 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: There is a big change in how it used to work and how it works now. To avoid race conditions modifications now happen exclusively under critical section. There is no longer concurrency allowed in modification. In fact, I didn't change the concurrent dictionary for plain dictionary with the only purpose of being as close to the original as possible to avoid introducing too extensive changes.

Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: The reason why `.TryRemoveInternal()` is because construction of frozen dictionaries must be done at the end of any modification procedure before leaving the lock. So actual modification operations must not create new dictionaries.

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 6 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: See https://github.com/ravendb/ravendb/pull/17942#discussion_r1441632156

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ResourceCache.cs]: I agree. Let's add a comment referencing RavenDB-19002

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ResourceCache.cs]: 1. So basically all modifications are done under the lock. That is this big change that you're talkin about, right?
2. I don't follow the need of introducing `TryRemoveInternal()`

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 3 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: Done already

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 5 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: It is a big change yes but it is not externally observable given the already racy semantics of the current code. Since we update dictionaries one at a time a reader may observe an unstable view (not a big deal for the usage pattern). 

Every remove and add actually now involves two stages. At stage 1 we do what we used to do in the method (now isolated in the internal one) and at stage 2 we create and update the references that are externally observable

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 9 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: It does, and it already existed in the current version. I will try to diminish the exposure a bit more. https://github.com/ravendb/ravendb/pull/17942#discussion_r1441767696

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 10]
```diff
@@ -86,44 +100,63 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
...
+            Debug.Assert(Monitor.IsEntered(this));
+
             if (_mappings.TryGetValue(resourceName, out ConcurrentSet<StringSegment> mappings))
             {
                 foreach (var mapping in mappings)
                 {
+                    // Careful here, the TryRemove method by its documentation would perform a (key,value) check instead of 
+                    // removing the item which has a matching key. This is an important departure of usual usage and has already
+                    // been the source of RavenDB-19002 
+                    // https://github.com/ravendb/ravendb/commit/34c9fcb5a111b352795f30ce24bd91d7a68bfe31
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ResourceCache.cs:139]: Let's apply the same comment to the above
```
if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
```

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 11]
```diff
@@ -36,24 +44,29 @@ public sealed class ResourceDetails
...
-        public IEnumerable<Task<TResource>> Values => _caseInsensitive.Values;
+        public IEnumerable<Task<TResource>> Values => _readonlyCaseInsensitive.Values;
 
-        public int Count => _caseInsensitive.Count;
+        public int Count => _readonlyCaseInsensitive.Count;
 
-        internal int DetailsCount => _resourceDetails.Count;
+        internal int DetailsCount => _readonlyResourceDetails.Count;
 
         public void Clear()
```
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs:53]: This should not be under a lock?

Comment (ayende) [reply src/Raven.Server/Documents/ResourceCache.cs:53]: I know that it is almost never called, but still

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 12]
```diff
@@ -86,44 +100,63 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
...
-            if (TryRemove(resourceName, resourceTask) == false)
-                return false;
+                // We need to refresh all of them. Since construction takes time, we will update all the references
+                // at the end of the execution in order to diminish the time that an unstable state can be observed.
+                // While this is not a big issue because the code already supports dealing with that, the smaller the
+                // time such inconsistencies can be observed, the better.
+                var readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                var readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                var readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
+
```
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs:119]: I guess we code wrap this in a Tuple, and replace that atomically, That would require additional indirection always, probably not a good idea, I'm assuming

File: src/Raven.Server/Documents/ResourceCache.cs [hunk 13]
```diff
@@ -306,6 +368,12 @@ public Task<TResource> Replace(string databaseName, Task<TResource> task)
                             _parent._caseSensitive.TryRemove(mapping, out Task<TResource> _);
                         }
                     }
+
+                    // We need to refresh all of them.
+                    _parent._readonlyCaseSensitive = _parent._caseSensitive.ToFrozenDictionary();
+                    _parent._readonlyCaseInsensitive = _parent._caseInsensitive.ToFrozenDictionary();
+                    _parent._readonlyResourceDetails = _parent._resourceDetails.ToFrozenDictionary();
+
```
Comment (ayende) [inline src/Raven.Server/Documents/ResourceCache.cs]: Shouldn't this be also done in the two steps method?
For that matter, can we wrap that in a method to ensure that this is not duplicated?

Comment (ppekrol) [reply src/Raven.Server/Documents/ResourceCache.cs]: @redknightlois ?

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 12 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/ResourceCache.cs:119]: @redknightlois  ?

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 13 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs:376]: What do you mean by the two-steps method? In this particular code, the update will be executed under the lock every time, regardless of what happens.

Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs]: Not a big deal, but yes. Fixed.

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 12 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs:119]: Not really  a big deal, the code itself handles this explicitly. There is no upside to do that and the code becomes much less legible.

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 11 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/ResourceCache.cs:53]: @redknightlois ?

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/ResourceCache.cs (hunk 11 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/ResourceCache.cs:53]: BTW:  Is `TryGetValue`  thread unsafe now? In case when you perform clean/remove in other thread,

Comment (redknightlois) [reply src/Raven.Server/Documents/ResourceCache.cs:53]: Those are readonly, worst case scenario you are asking to the wrong dictionary.
